### PR TITLE
Return correct user level for users for levels

### PR DESCRIPTION
**Pull request by maureensturgeon**
There is a bug where duplicated user levels are created for a user on a level, this issue has been discovered by a teacher when they realized that the state of a progress bubble on their teacher dashboard for the student was not the same as the state of the progress bubble on the level page for that student. Here we've added some sorting logic to ensure the correct user level is returned for the teacher dashboard while we wait for the full fix from infra.

## Testing story
Tested that with duplicated user levels the correct one returns for the progress table. Also added unit tests

## Follow-up work
Full fix: https://codedotorg.atlassian.net/browse/INF-571

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/user.rb [hunk 1]
```diff
@@ -1081,10 +1081,21 @@ def self.authenticate_with_section_and_secret_picture(section:, params:)
...
 
+  # There is a bug (fix: https://codedotorg.atlassian.net/browse/INF-571) where some users have
+  # duplicate user levels for the same level. To ensure that we return the relevant user level for
+  # each level and not one of the duplicates, the list is first sorted so that the
+  # most relevant user levels are at the end. The list is then indexed by level ID, which will
+  # pick up the last matching user level in the list.
+  def self.index_user_levels_by_level_id(user_levels)
+    # The correct user level is the one most recently updated or the first created
+    relevant_user_levels_last = user_levels.sort {|a, b| [a.updated_at, b.id] <=> [b.updated_at, a.id]}
+    relevant_user_levels_last.index_by(&:level_id)
```
Comment (maureensturgeon) [inline dashboard/app/models/user.rb:1093]: I could replace these two lines with a custom index_by where I do the comparison while iterating to build the resulting list, which would probably be more efficient, but this is more concise. Let me know if you have opinions.

**Review by maureensturgeon — COMMENTED** [review]
This should be a temporary patch, do we need unit tests?

**Comment by maureensturgeon** [comment]
> This should be a temporary patch, do we need unit tests?

I added a couple unit tests!

**Review by jamescodeorg — APPROVED** [review]
Thanks for jumping on this!  Could you open a ticket to revert this fix when the underlying issue is fixed?

**Pull request merged**